pkg/controller/networkpolicy/mutate.go

```diff
@@ -0,0 +1,201 @@
+	var result *metav1.Status
+	var msg string
+	var patch []byte
+	allowed := false
+	patchType := admv1.PatchTypeJSONPatch
+
+	op := ar.Request.Operation
+	curRaw := ar.Request.Object.Raw
+	oldRaw := ar.Request.OldObject.Raw
+
+	switch ar.Request.Kind.Kind {
+	case "ClusterNetworkPolicy":
+		klog.V(2).Info("Mutating Antrea ClusterNetworkPolicy CRD")
+		var curACNP, oldACNP secv1alpha1.ClusterNetworkPolicy
+		if curRaw != nil {
+			if err := json.Unmarshal(curRaw, &curACNP); err != nil {
+				klog.Errorf("Error de-serializing current Antrea ClusterNetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		if oldRaw != nil {
+			if err := json.Unmarshal(oldRaw, &oldACNP); err != nil {
+				klog.Errorf("Error de-serializing old Antrea ClusterNetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		msg, allowed, patch = m.mutateAntreaPolicyRuleName(op, curACNP.Spec.Ingress, curACNP.Spec.Egress)
+	case "NetworkPolicy":
+		klog.V(2).Info("Mutating Antrea NetworkPolicy CRD")
+		var curANP, oldANP secv1alpha1.NetworkPolicy
+		if curRaw != nil {
+			if err := json.Unmarshal(curRaw, &curANP); err != nil {
+				klog.Errorf("Error de-serializing current Antrea NetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		if oldRaw != nil {
+			if err := json.Unmarshal(oldRaw, &oldANP); err != nil {
+				klog.Errorf("Error de-serializing old Antrea NetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		msg, allowed, patch = m.mutateAntreaPolicyRuleName(op, curANP.Spec.Ingress, curANP.Spec.Egress)
+	}
+
+	if msg != "" {
+		result = &metav1.Status{
+			Message: msg,
+		}
+	}
+	return &admv1.AdmissionResponse{
+		Allowed:   allowed,
+		Result:    result,
+		PatchType: &patchType,
+		Patch:     patch,
+	}
+}
+
+// mutateAntreaPolicyRuleName mutates names of rules of an Antrea NetworkPolicy CRD.
+// If users didn't specify the name of an ingress or egress rule,
+// mutateAntreaPolicyRuleName will auto-generate a name for this rule.
+func (m *NetworkPolicyMutator) mutateAntreaPolicyRuleName(op admv1.Operation, ingress, egress []secv1alpha1.Rule) (string, bool, []byte) {
+	allowed := true
+	reason := ""
+	var patch []byte
+	switch op {
+	case admv1.Create, admv1.Update:
+		ruleNames := sets.NewString()
+		for _, rule := range append(ingress, egress...) {
+			ruleNames.Insert(rule.Name)
+		}
+
+		ingressPath, ingressValue := generateName(ruleNames, "ingress", ingress)
+		egressPath, egressValue := generateName(ruleNames, "egress", egress)
+
+		genPatch, err := createReplacePatch(append(ingressPath, egressPath...), append(ingressValue, egressValue...))
+		if err != nil {
+			allowed = false
+			reason = "unable to generate mutating patch"
+			break
+		}
+		patch = genPatch
+
+	case admv1.Delete:
+		// Delete of Antrea Policies have no mutation
+		allowed = true
+	}
+	return reason, allowed, patch
+}
+
+// generateName generate unique rule name and return a list of jsonPath and a list generated name
```

Comment:
call the function `generateRuleNames`?

and use this function comment: "generateRuleNames generates unique rule names and returns a list of json paths and the corresponding list of generated names"

Reply:
Sure thx.